compiler/rustc_transmute/src/layout/tree.rs

```diff
@@ -187,15 +187,15 @@ pub(crate) mod rustc {
     pub(crate) enum Err {
         /// The layout of the type is unspecified.
         Unspecified,
-        /// This error will be surfaced elsewhere by rustc, so don't surface it.
-        Unknown,
+        Unknown(ErrorGuaranteed),
         TypeError(ErrorGuaranteed),
     }
 
     impl<'tcx> From<LayoutError<'tcx>> for Err {
         fn from(err: LayoutError<'tcx>) -> Self {
+            let guar = rustc_middle::ty::tls::expect_compilation_to_fail();
```

Comment:
A layout error does not imply that compilation is going to fail. This can happen when the type is sufficiently polymorphic, such as `&T` where `T: ?Sized`.

Reply:
This would ICE, for example, when we relax this requirement:

https://github.com/rust-lang/rust/blob/80a2ec49a4ffb7a351c41c8db14711297324b587/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs#L778-L780

Reply:
If I'm understanding correctly, you're saying that a layout error _may or may not_ cause compilation to fail? If so, I can change it to `Unknown(Option<ErrorGuaranteed>),` and use `if let Some(guar) = with(|tcx| tcx.sess.is_compilation_going_to_fail())` to decide what to return in this `From` impl

Reply:
No, I don't really think that relying on global state such as that for trait solver behavior is a good idea.

Reply:
Gotcha, I can undo the changes in this section then